### PR TITLE
test(api): e2e core library for rds tests

### PR DESCRIPTION
**Pull request by sundersc**
#### Description of changes

E2E Core library for RDS tests. This includes the below helpers.

- Create a RDS Instance
- Add external IP of current machine running the test to security group
- Remove the IP from the security group
- Delete a RDS Instance
- A sample test to show how to use the above methods

#### Description of how you validated changes
- E2E Test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-e2e-core/src/utils/rds.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      throw new Error("RDS Instance details are missing.");
+    }
+
+    return {
+      endpoint: dbInstance.Endpoint.Address,
+      port: dbInstance.Endpoint.Port,
+      dbName: dbInstance.DBName,
+    };
+  } catch (error) {
+    console.log(error);
```
Comment (alharris-at) [inline packages/amplify-e2e-core/src/utils/rds.ts:67]: Probably want `console.error` here.

File: packages/amplify-e2e-core/src/utils/rds.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  waitUntilDBInstanceDeleted 
+} from "@aws-sdk/client-rds";
+import { EC2Client, AuthorizeSecurityGroupIngressCommand, RevokeSecurityGroupIngressCommand } from '@aws-sdk/client-ec2';
+import { knex } from 'knex';
+
+const DEFAULT_DB_INSTANCE_TYPE = "db.m5.large";
+const DEFAULT_DB_STORAGE = 8;
+const DEFAULT_SECURITY_GROUP = "default";
+
+export const createRDSInstance = async (config: {
```
Comment (alharris-at) [inline packages/amplify-e2e-core/src/utils/rds.ts:15]: Could you add docs to these new utils. No time like the present to improve the state of our docs 👯

Comment (alharris-at) [reply packages/amplify-e2e-core/src/utils/rds.ts:15]: Could you also add response types to the declaration?

File: packages/amplify-e2e-core/src/utils/rds.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  const client = new RDSClient({ region });
+  const params = {
+    "DBInstanceIdentifier": identifier,
+    "SkipFinalSnapshot": true,
+  };
+  const command = new DeleteDBInstanceCommand(params);
+  try {
+    await client.send(command);
+    
+    // TODO: Revisit the below logic for waitUntilDBInstanceDeleted.
```
Comment (alharris-at) [inline packages/amplify-e2e-core/src/utils/rds.ts:82]: When are we revisiting this? What's blocking us getting this in place today?

**Review by alharris-at — COMMENTED** [review]
Nice! Just a few comments about making sure our utilities are nicely documented, etc.

File: packages/amplify-e2e-core/src/utils/rds.ts (hunk 2 above)
Comment (sundersc) [reply packages/amplify-e2e-core/src/utils/rds.ts:20]: Added docs for the new methods.

File: packages/amplify-e2e-core/src/utils/rds.ts (hunk 3 above)
Comment (sundersc) [reply packages/amplify-e2e-core/src/utils/rds.ts:92]: I see some issues with the way SDK method 'waitUntilDBInstanceDeleted' works. It polls for the instance status at regular intervals and returns once the status is 'DELETED'. But the problem is once we trigger the instance deletion, it takes few minutes for the instance to be actually deleted. SDK returns 'DELETED' status only for few seconds after instance deletion. After that, it returns 'NO SUCH INSTANCE' error. I can reduce the interval to a lower value and get it to work most of the times but still there are changes of flaky test.

With current implementation, we trigger the deletion and exit the test. If test doesn't fail at this step, the instance will be deleted few minutes (usually 2-3 minutes) after the test run.

**Review by alharris-at — APPROVED** [review]

File: packages/amplify-e2e-tests/src/__tests__/rds-v2.test.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    // Get the public IP of the machine running the test
+    const url = "http://api.ipify.org/";
+    const response = await axios(url);
+    publicIpCidr = `${response.data.trim()}/32`;
+  });
+
+  it("create database, setup initial tables and delete database", async () => {
+    // This test performs the below
+    // 1. Create a RDS Instance
+    // 2. Add the external IP address of the current machine to security group inbound rule to allow public access
```
Comment (phani-srikar) [inline packages/amplify-e2e-tests/src/__tests__/rds-v2.test.ts:19]: If for some reason these instances are not cleaned up, allowing public access might raise security alarms?

**Review by phani-srikar — APPROVED** [review]

File: packages/amplify-e2e-tests/src/__tests__/rds-v2.test.ts (hunk 4 above)
Comment (sundersc) [reply packages/amplify-e2e-tests/src/__tests__/rds-v2.test.ts:19]: We delete the security group inbound rule at the end of the test. Even for some reason if it fails to delete the rule, we allow access to one specific IP so the risk is very low. Also, we run the test in a separate account and there is no sensitive data on this account.

**Pull request merged**